autorel/utils/docker_cli.py

```diff
@@ -0,0 +1,150 @@
+        """
+            Executes a command in a running docker container
+            and checks for the execution status
+        """
+        cmd = 'bash -c \"{0}\"'.format(command_)
+        self._logger.debug("Executing in container ({0}) : {1}"
+            .format(self._container_id,cmd))
+        exec_instance = self._cli_command("exec_create",
+                                          response=True,
+                                          container=self._container_id,
+                                          cmd=cmd
+                                          )
+        exec_id = exec_instance.get('Id')
+        self._cli_command("exec_start",
+                          exec_id
+                          )
+        exec_status = self._cli_command("exec_inspect",
+                                        exec_id,
+                                        response=True
+                                        )
+        if exec_status["ExitCode"] != 0:
+            raise CommandFailure(response)
+
+
+    def _setup(self, image, source_directory):
+        """
+            Sets up the building environment by pulling the
+            image and starting the docker container
+        """
+        ## Pull the docker image
+        self._logger.info("Pulling the {0} docker image"
+            .format(image))
+        self._cli_command("pull",
+                          repository=image
+                          )
+        self._logger.info("Image successfuly pulled")
+
+        ## Create a container in daemon mode
+        self._logger.info("Running a container using the image")
+        host_config = self._cli_command("create_host_config",
+                                        response=True,
+                                        binds={
+                                            source_directory : {
+                                                   'bind': '/home/',
+                                                   'mode': 'rw'
+                                               }
+                                            }
+                                        )
+        container = self._cli_command("create_container",
+                                      response=True,
+                                      image=image,
+                                      detach=True,
+                                      command="tail -f /dev/null",
+                                      volumes=["/home/"],
+                                      host_config =host_config
+                                      )
+        self._container_id = container.get("Id")
+        # Start the container
+        self._cli_command("start",
+                          response=True,
+                          container=self._container_id)
+        self._running = True
+        self._logger.info("Container ({0}) is successfuly running"
+            .format(self._container_id))
+
+    def run(self, builder):
```

Comment:
I created these builder classes for mainly two reasons:
1) Keeping of commands to run
2) Chain results of sequential build processes
Using these, I was able to control the results in a much better way. Like for `DebianPackageBuilder` its result output is `DebianSourcePackage` (https://github.com/black-perl/syslog-ng-autorel/blob/develop/autorel/builders/debian_source_builder.py#L10 ) and then I can easily pass this to next building process.
If I pass the commands list, I can still return this; but in that case I need different implementations of `Docker::run` for each such building process. 
So, how we gonna deal with this then ?